### PR TITLE
downside support & generalize premium

### PR DESCRIPTION
**Pull request by DonDuala**
Preparing for Downside support & generalizing premium assets. Want this ready to go once we finish https://github.com/Dual-Finance/dual-labs-dip/issues/122

### CODE REVIEW

**Comment by DonDuala** [comment]
Fixes https://github.com/Dual-Finance/risk-manager/issues/56

File: src/utils.ts [hunk 1]
```diff
@@ -61,9 +61,11 @@ export function parseDipState(buf: Buffer) {
   const vaultSplBump = Number(buf.readUInt8(121));
   const optionMint = new PublicKey(buf.slice(122, 154));
   const optionBump = Number(buf.readUInt8(154));
-  const vaultUsdc = new PublicKey(buf.slice(155, 187));
-  const vaultUsdcBump = Number(buf.readUInt8(187));
-  const usdcMint = new PublicKey(buf.slice(188, 220));
+  const vaultPremium = new PublicKey(buf.slice(155, 187));
+  const vaultPremiumBump = Number(buf.readUInt8(187));
+  const premiumMint = new PublicKey(buf.slice(188, 220));
+  const type = String(buf.slice(188, 220));
```
Comment (brittcyr) [inline src/utils.ts:67]: This parser handles
https://github.com/Dual-Finance/dual-labs-dip/blob/master/programs/dual-market/src/common.rs

**Review by brittcyr — COMMENTED** [review]
Do we need a type and (premium asset + splMint)? Probably best to just remove type since we want to be able to support multiple premium assets

**Comment by DonDuala** [comment]
> 

The way we will calculate delta/gamma and manage the position will be in terms of call/put. Its fine to infer them from base/quote assets but i didnt see any current construction using those. It might get confusing to switch splMInt and premiumMInt to base & quote but can do if you think its best. The DIPs will have the type property so id prefer to do any inference of type elsewhere

**Comment by brittcyr** [comment]
Why do the DIPs need a type property? That was originally added in the risk manager before we had to generalize the DIP contract to arbitrary premium mints.
There is no type property in the contract, so should switch to base/quote here to match the contract.

Type is too general of a word, probably best to rename it direction or something else that explains what we are using it for.

File: src/router.ts [hunk 2]
```diff
@@ -170,11 +174,10 @@ export class Router {
     const balance = await connection.getTokenAccountBalance(mmOptionAccount);
 
     this.dips[this.dip_to_string(expirationSec, strike)] = {
-      splToken: splMintToToken(splMint),
-      premiumAsset: "USD",
+      baseAsset: splMintToToken(baseMint),
+      quoteAsset: splMintToToken(quoteMint),
       expirationMs: expirationSec * 1_000,
       strike: strike,
```
Comment (DonDuala) [inline src/router.ts:180]: Should I use adjustStrike here too? Do we risk collisions of optionMints if we create a DIP of the same strike & expiration but different quote/base (call & put)?

File: src/utils.ts [hunk 3]
```diff
@@ -54,47 +55,49 @@ export function readBigUInt64LE(buffer: Buffer, offset = 0) {
...
   const vaultSpl = new PublicKey(buf.slice(89, 121));
   const vaultSplBump = Number(buf.readUInt8(121));
   const optionMint = new PublicKey(buf.slice(122, 154));
   const optionBump = Number(buf.readUInt8(154));
-  const vaultUsdc = new PublicKey(buf.slice(155, 187));
-  const vaultUsdcBump = Number(buf.readUInt8(187));
-  const usdcMint = new PublicKey(buf.slice(188, 220));
+  const vaultPremium = new PublicKey(buf.slice(155, 187));
+  const vaultPremiumBump = Number(buf.readUInt8(187));
+  const quoteMint = new PublicKey(buf.slice(188, 220));
```
Comment (DonDuala) [inline src/utils.ts:67]: Not sure if some of these are redundant?

File: src/poller.ts [hunk 4]
```diff
@@ -47,7 +44,7 @@ export class Poller {
       // @ts-ignore
       const new_amount = parseTokenAccount(accountInfo.data).amount.toNumber();
       let decimals = 6;
-      switch(this.splToken) {
+      switch(this.baseAsset) {
         // BTC
```
Comment (brittcyr) [inline src/poller.ts:48]: downside can have USDC here. I think that is 6 decimals.

File: src/utils.ts (hunk 3 above)
Comment (brittcyr) [reply src/utils.ts:67]: quoteMint and quoteAsset? They are reading the same thing, so should only need one.

File: src/utils.ts [hunk 5]
```diff
@@ -137,17 +134,17 @@ export function timeSinceMidDay() {
...
-  if (splMint.toBase58() == soBTCPk.toBase58()) {
+  if (baseMint.toBase58() == soBTCPk.toBase58()) {
     return "BTC";
   }
-  if (splMint.toBase58() == soETHPk.toBase58()) {
+  if (baseMint.toBase58() == soETHPk.toBase58()) {
     return "ETH";
   }
-  if (splMint.toBase58() == usdcMintPk.toBase58()) {
+  if (baseMint.toBase58() == usdcMintPk.toBase58()) {
```
Comment (brittcyr) [inline src/utils.ts:147]: USDH also?

File: src/utils.ts [hunk 6]
```diff
@@ -179,27 +176,53 @@ export function tokenToPythSymbol(token: string) {
...
+    if (dip_deposit.baseAsset == premium) {
+      return "put";
+    }
+  }
+  return 'undefined';
+}
+
+export function getDIPDirection(dip_deposit: DIPDeposit) {
+  for (let premium of premiumAssets){
+    if (dip_deposit.quoteAsset == premium) {
```
Comment (brittcyr) [inline src/utils.ts:208]: is this functionally different from getDIPType?

File: src/utils.ts [hunk 7]
```diff
@@ -179,27 +176,53 @@ export function tokenToPythSymbol(token: string) {
...
+  return 'undefined';
+}
+
+export function adjustStrike(dip_deposit: DIPDeposit) {
+  for (let premium of premiumAssets){
+    if (dip_deposit.quoteAsset == premium) {
+      return dip_deposit.strike;
+    }
+    if (dip_deposit.baseAsset == premium) {
+      return 1 / dip_deposit.strike;
```
Comment (brittcyr) [inline src/utils.ts:224]: Not sure if the units work out here. Is strike based on what we use for the scalper, or what is in the DIPState on chain? The one on chain is integers only, so it is atoms of USDC per full token.
Either way, at least leave a TODO to change the property strike to include units.

File: src/utils.ts [hunk 8]
```diff
@@ -179,27 +176,53 @@ export function tokenToPythSymbol(token: string) {
...
     }
   }
   return 0.0;
 }
 
-export function tokenType(type: string) {
-  if (type == "Upside") {
-    return "call";
+export function getDIPType(dip_deposit: DIPDeposit) {
+  for (let premium of premiumAssets){
```
Comment (brittcyr) [inline src/utils.ts:195]: using a for loop to iterate over premium will work, but is unnecessary. Fine for now, but something to clean up later.

File: src/utils.ts [hunk 9]
```diff
@@ -179,27 +176,53 @@ export function tokenToPythSymbol(token: string) {
...
+    if (dip_deposit.quoteAsset == premium) {
+      return "Upside";
+    }
+    if (dip_deposit.baseAsset == premium) {
+      return "Downside";
+    }
   }
-  if (type == "Downside") {
-    return "put";
+  return 'undefined';
```
Comment (brittcyr) [inline src/utils.ts:215]: This should log and possibly throw an exception

File: src/router.ts (hunk 2 above)
Comment (brittcyr) [reply src/router.ts:180]: To avoid collision, change dip_to_string 
I think that is only used as a key for this.

File: src/utils.ts (hunk 6 above)
Comment (DonDuala) [reply src/utils.ts:208]: Just return Upside or Downside conform to this from the API
SPL_TOKEN,PREMIUM_ASSET,EXPIRATION,STRIKE,TYPE,POOL,STYLE,SETTLEMENT
Example:
BTC,USDC,2022-07-08,22000,UPSIDE,DIP,A,P

File: src/utils.ts (hunk 3 above)
Comment (DonDuala) [reply src/utils.ts:67]: vaultPremium & quoteMint/ quoteAsset

File: src/utils.ts (hunk 6 above)
Comment (DonDuala) [reply src/utils.ts:208]: Ah this will break threshold price, updating

File: src/utils.ts (hunk 3 above)
Comment (brittcyr) [reply src/utils.ts:67]: Most of the stuff in here isnt actually needed, was just included for completeness of parsing the program account.

File: src/utils.ts (hunk 6 above)
Comment (brittcyr) [reply src/utils.ts:208]: Should have the logic of determining direction in one just one place. Will help with code maintainability

Comment (DonDuala) [reply src/utils.ts:208]: Should this sit someplace in the DIP repo? DIP state could be read inclusive of getDIPType & adjustStrike?

**Pull request closed without merging**